messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/MessagingExtension.java

```diff
@@ -199,14 +207,11 @@ public void initialize(ExtensionContext context) {
             deployedServer.registerSubModel(new JMSTopicDefinition(true, registerRuntimeOnly));
             deployedServer.registerSubModel(PooledConnectionFactoryDefinition.DEPLOYMENT_INSTANCE);
         }
-
-        if (context.isRegisterTransformers()) {
-            MessagingSubsystemRootResourceDefinition.registerTransformers(subsystemRegistration);
-        }
     }
 
     public void initializeParsers(ExtensionParsingContext context) {
         context.setSubsystemXmlMapping(SUBSYSTEM_NAME, MessagingSubsystemParser_1_0.NAMESPACE, MessagingSubsystemParser_1_0::new);
         context.setSubsystemXmlMapping(SUBSYSTEM_NAME, MessagingSubsystemParser_2_0.NAMESPACE, MessagingSubsystemParser_2_0::new);
+        context.setSubsystemXmlMapping(SUBSYSTEM_NAME, MessagingSubsystemParser_3_0.NAMESPACE, MessagingSubsystemParser_3_0::new);
```

Comment:
Please use new MessagingSubsystemParser_3_0(), but keep the method reference for the other versions.

This is a bit of black art. When a Supplier is not used, ExtensionRegistry is able to detect use of a PersistentResourceXMLParser and it then triggers a call to getParserDescription(). The object being to have that call happen during extension initialization, which is part of the parallelized part of boot, rather than having it be deferred until parsing, which is single-threaded. Since a getParserDescription() impl typically results in a low of loading of classes from the extension impl, execution of all the static initialization therein, etc, there's a pretty decent boot time gain from getting it out of the single threaded parts of boot.

Using a Supplier for the legacy versions prevents pointless execution of code for those. There's a small metaspace cost though. Tradeoffs. :)


(One of my goals for next-gen declarative subsystem dev is to make getting this stuff right automagic, as this is definitely black art now. I added the ExtensionRegistry trick late in the WF 11 cycle when we were looking for small boot time gains.)

Reply:
note for later, maybe the API should distinguish between registering the subsystem XML mapping for the current version (and load an instance) vs. legacy XML mappings (where a supplier would prevent unnecessary code execution).

Reply:
fixed (by amending the WFLY-9407 commit)

Reply:
@jmesnil My dream is you just register your model and we generate the current parser. Specifically registered parsers would be for old versions only.